docs/sdg/sdg-refactor.md

```diff
@@ -0,0 +1,147 @@
+    - 1.3. They have raw seed content.  They want to run SDG only.
+        - 1.3.1. They also want to see the inputs to SDG that get extracted from the raw seed content (i.e., a set of seed context/question/answer tuples and with document chunks for the knowledge if any).
+        - 1.3.2. Alternatively, maybe they only want to see the inputs to SDG -- they don't actually want to run SDG.
+    - 1.4. They have processed seed content.  They want to run the full pipeline including SDG and model training and evaluation.
+    - 1.5. They have processed seed content.  They want to run SDG and then evaluate an existing model on the outputs of that SDG.
+    - 1.6. They have processed seed content.  They want to run SDG only.
+2. User installs the SDG library only.  They want to invoke any of the following as a library call from code they write:
+    - 2.1. They have raw seed content.  They want to run SDG only without any postprocessing.
+    - 2.2. They have processed seed content.  They want to run SDG only without any postprocessing.
+
+If I understand the latest guidance from our product management, the flows that our users want us to support here are 1.1., 1.2., 1.3, 1.3.1, and 1.6.  In an earlier draft of this proposal, I had said that I thought product management also wanted 2.2, but the latest guidance doesn't seem consistent with that understanding.  I am still not sure, so more clarification would be helpful.
+
+## Question 2: What should the commands be in the CLI?
+
+One way to support both 1.3.1 and 1.3.2 would be to have separate CLI commands for the preprocessing, core SDG, and postprocessing step .  Alternatively, a single CLI command that does all of these and also saves the outputs of preprocessing to disk would support 1.3.1 but *not* 1.3.2.  Even if we only want to support 1.3.1, having separate CLI commands for each step might be desirable because it is just more intuitive that if a user wants to save the outputs of preprocessing to disk to have a command to do that instead of having it be a "side effect" of an omnibus SDG command.  Here is a rough outline of what the separate commands would be:
+
+- `ilab data prep` would handle all the preprocessing (the first three bullets in the Context section above, plus any additional preprocessing we add in the future).
+- `ilab data generate` would take as input some data in the same format that `ilab data prep` produces and would run the core synthetic data generation *only*.  Note that this is a breaking change from the current behavior of `ilab data generate`, but that may be acceptable because the user base is still small.
+- `ilab data process` would take as input some data in the same format that `ilab data generate` produces and would run the postprocessing (the last two bullets in the Context section above, plus any additional postprocessing we add in the future).
```

Comment:
`process` -> `postprocess`? 

Can we find a better name? As you mention, there may be other types of processing or postprocessing added in the future. What would be a name that indicates the purpose of this operation?

Reply:
@cdoern , you proposed this name as I recall.  I agree with Anastas that it is kind of generic.  I guess I like `postprocess` a little better.  At one point we had called it "mix" or "mixing" but I think it is already broader than just mixing and will probably get even broader in the future.

Reply:
sure, post process sounds good to me!

Reply:
I lean towards `ilab data mix` as well as I don't think ambiguity or generic-ness helps us much here.

In mixing, we already have the concept of the 'recipe' yaml files which serve as formulas by which to sample and mix data. To that end, I think we should have the mixing command be cohesive and self-descriptive, if we want to empower users to tweak or write their own recipe files. This way, at least the flags we expose in the `ilab data mix` command make sense contextually.

Also, I'd argue that if any clear other candidates emerge as post processing steps we'd want to do, if they are different enough to not be lumped into mixing, that'd probably justify having a separate command if we do want to make it configurable to users.

Reply:
I have been hearing conflicting reports about whether all of the postprocessing we do really is data mixing.  If we have a command like `ilab data mix` and we want to do any postprocessing that is not really data mixing then we have two unappealing options:

1. Add that as part of `ilab data mix` and accept that the name is not really very accurate.
2. Add another command that runs before or after `ilab data mix` or possibly one command that runs before and another that runs after for pre-mix postprocessing and post-mix postprocessing.

My main concern is that we will actually pick option two because we have a long track record of not updating our names for concepts when those names no longer apply.  For example, the qna.yaml files for skills have `question` and `answer` pairs because that is the format but really most skills (unlike knowledge) are not really about answering questions but rather about fulfilling a request and providing a response.  Renaming things is a hassle so we stick with the old names.  This suggests that going forward we should try to use more generic names that will be broad enough to cover not only what we do but also what we expect to do.

Reply:
One of the goals of the domain modeling workshop was to encourage folks to think about names from a more user-centric perspective, which would mean trying to choose more accurate names at least. This seems like a good time to spend some time picking good names, like how we spent time trying to name our bounded contexts at the workshop.